environment.go

```diff
@@ -16,10 +16,120 @@ package esc
 
 import (
 	"encoding/json"
+	"errors"
+	"fmt"
 
 	"github.com/pulumi/esc/schema"
 )
 
+const AnonymousEnvironmentName = "<yaml>"
+
+type ExecContext struct {
+	rootEvironment string
+	values         map[string]Value
+}
+
+type copier struct {
+	memo map[*Value]*Value
+}
+
+func newCopier() copier {
+	return copier{memo: map[*Value]*Value{}}
+}
+
+func (c copier) copy(v *Value) *Value {
+	if v == nil {
+		return nil
+	}
+
+	if copy, ok := c.memo[v]; ok {
+		return copy
+	}
+
+	copy := &Value{}
+	c.memo[v] = copy
+
+	var nv any
+	switch vr := v.Value.(type) {
+	case []*Value:
+		a := make([]*Value, len(vr))
+		for i, v := range vr {
+			a[i] = c.copy(v)
+		}
+		nv = a
+	case map[string]*Value:
+		m := make(map[string]*Value, len(vr))
+		for k, v := range vr {
+			m[k] = c.copy(v)
+		}
+		nv = m
+	default:
+		nv = vr
+	}
+
+	*copy = Value{
+		Value: nv,
+	}
+	return copy
+}
+
+func copyContext(context map[string]Value) map[string]Value {
+	newContext := make(map[string]Value)
+	for key, v := range context {
+		value := v
+		copy := newCopier().copy(&value)
+		newContext[key] = *copy
+	}
+	return newContext
+}
+
+func (ec *ExecContext) Copy() *ExecContext {
+	return &ExecContext{
+		values:         copyContext(ec.values),
+		rootEvironment: ec.rootEvironment,
+	}
+}
+
+func (ec *ExecContext) Values(envName string) map[string]Value {
+	context := ec.values
+	context["currentEnvironment"] = NewValue(map[string]Value{
+		"name": NewValue(envName),
+	})
+
+	if ec.rootEvironment == AnonymousEnvironmentName || ec.rootEvironment == "" {
+		ec.rootEvironment = envName
+	}
+
+	context["rootEnvironment"] = NewValue(map[string]Value{
+		"name": NewValue(ec.rootEvironment),
+	})
+
+	return context
+}
+
+var ErrForbiddenContextkey = errors.New("forbidden context key")
```

Comment:
nit: we generally use "reserved" for keys that the user is not allowed to set

```suggestion
var ErrReservedContextkey = errors.New("reserved context key")
```

Reply:
done